### PR TITLE
Fix parser + cleanups

### PR DESCRIPTION
**Pull request by marrony**
That PR fix a issue found in Json parser when an empty object literal was passed in like `{"@obj": {}}`. It also include some minor internal refactorings to make the types more consistent.

### CODE REVIEW

File: FaunaDB.Client/Collections/ImmutableDictionary.cs [hunk 1]
```diff
@@ -1,81 +1,176 @@
...
+using System.Collections.Specialized;
 
 namespace FaunaDB.Collections
 {
+    class ImmutableDictionary<TKey, TValue> : IReadOnlyDictionary<TKey, TValue>
+    {
+        public static readonly ImmutableDictionary<TKey, TValue> Empty =
+            new ImmutableDictionary<TKey, TValue>();
+
+        readonly OrderedDictionary dictionary = new OrderedDictionary();
```
Comment (erickpintor) [inline FaunaDB.Client/Collections/ImmutableDictionary.cs]: Did `OrderedDictionary` already existed in the standard library?

File: FaunaDB.Client/Types/ObjectV.cs [hunk 2]
```diff
@@ -14,13 +14,14 @@ namespace FaunaDB.Types
...
+        public static readonly ObjectV Empty =
+            new ObjectV(ImmutableDictionary<string, Value>.Empty);
 
         public IReadOnlyDictionary<string, Value> Value { get; }
 
-        public ObjectV() : this(Empty.Value) { }
+        internal ObjectV() : this(Empty.Value) { }
 
-        public ObjectV(IReadOnlyDictionary<string, Value> value)
+        internal ObjectV(IReadOnlyDictionary<string, Value> value)
```
Comment (erickpintor) [inline FaunaDB.Client/Types/ObjectV.cs]: Did we drop all public constructors for `ObjectV`?

File: FaunaDB.Client/Collections/ImmutableDictionary.cs (hunk 1 above)
Comment (marrony) [inline FaunaDB.Client/Collections/ImmutableDictionary.cs]: Yes, but it's not generic

File: FaunaDB.Client/Types/ObjectV.cs (hunk 2 above)
Comment (marrony) [inline FaunaDB.Client/Types/ObjectV.cs]: Yes, `ObjectV` instances can only be created through `Obj()` method

Comment (marrony) [inline FaunaDB.Client/Types/ObjectV.cs]: Ops, `ObjectV` is only created when parsing the response, `Obj()` create unwrapped objects

**Pull request merged**